server/src/main/java/org/opensearch/gateway/remote/RemoteGlobalMetadataManager.java

```diff
@@ -277,29 +280,37 @@ Metadata getGlobalMetadata(String clusterUUID, ClusterMetadataManifest clusterMe
         }
     }
 
-    Map<String, Metadata.Custom> getUpdatedCustoms(ClusterState currentState, ClusterState previousState) {
-        if (Metadata.isCustomMetadataEqual(previousState.metadata(), currentState.metadata())) {
-            return new HashMap<>();
-        }
-        Map<String, Metadata.Custom> updatedCustom = new HashMap<>();
-        Set<String> currentCustoms = new HashSet<>(currentState.metadata().customs().keySet());
-        for (Map.Entry<String, Metadata.Custom> cursor : previousState.metadata().customs().entrySet()) {
-            if (cursor.getValue().context().contains(Metadata.XContentContext.GATEWAY)) {
-                if (currentCustoms.contains(cursor.getKey())
-                    && !cursor.getValue().equals(currentState.metadata().custom(cursor.getKey()))) {
-                    // If the custom metadata is updated, we need to upload the new version.
-                    updatedCustom.put(cursor.getKey(), currentState.metadata().custom(cursor.getKey()));
-                }
-                currentCustoms.remove(cursor.getKey());
-            }
+    DiffableUtils.MapDiff<String, Metadata.Custom, Map<String, Metadata.Custom>> getCustomsDiff(
+        ClusterState currentState,
+        ClusterState previousState,
+        boolean firstUploadForSplitGlobalMetadata,
+        boolean includeEphemeral
+    ) {
+        if (firstUploadForSplitGlobalMetadata) {
+            // For first split global metadata upload, we want to upload all customs
+            return DiffableUtils.diff(
+                Collections.emptyMap(),
+                filterCustoms(currentState.metadata().customs(), includeEphemeral),
+                DiffableUtils.getStringKeySerializer(),
+                NonDiffableValueSerializer.getAbstractInstance()
+            );
         }
-        for (String custom : currentCustoms) {
-            Metadata.Custom cursor = currentState.metadata().custom(custom);
-            if (cursor.context().contains(Metadata.XContentContext.GATEWAY)) {
-                updatedCustom.put(custom, cursor);
-            }
+        return DiffableUtils.diff(
+            filterCustoms(previousState.metadata().customs(), includeEphemeral),
+            filterCustoms(currentState.metadata().customs(), includeEphemeral),
+            DiffableUtils.getStringKeySerializer(),
+            NonDiffableValueSerializer.getAbstractInstance()
```

Comment:
Would this be correct if we enabled publication in last cluster state. We would have only uploaded the customs required to be persisted. But after publication is enabled we need to upload previously not uploaded customs as well.

Reply:
For now this would work because when publication is enabled, the new cluster manager will publish the full state with all ephemeral entities after the cluster is up.
But if we plan to do incremental remote upload even on first upload from new cluster manager then we need to handle this case.